cluster.go

```diff
@@ -161,34 +162,67 @@ func (p SchedulerParams) MergeConfig(base *schedulerv1beta1.KubeSchedulerConfigu
 // ProxyParams is a set of extra parameters for kube-proxy.
 type ProxyParams struct {
 	ServiceParams `json:",inline"`
-	Mode          ProxyMode `json:"mode"`
+	Config        *unstructured.Unstructured `json:"config,omitempty"`
 }
 
 // GetMode returns the proxy mode.
-func (p ProxyParams) GetMode() ProxyMode {
-	if len(p.Mode) == 0 {
-		return ProxyModeIPVS
+func (p ProxyParams) GetMode() string {
+	mode := p.Config.UnstructuredContent()["Mode"].(string)
+	if len(mode) == 0 {
+		return string(ProxyModeIPVS)
 	}
-	return p.Mode
+	return mode
+}
+
+// MergeConfig merges the input struct with `base`.
+func (p ProxyParams) MergeConfig(base *proxyv1alpha1.KubeProxyConfiguration) (*proxyv1alpha1.KubeProxyConfiguration, error) {
+	// FOR IMPLEMENTORS.
+	// DO NOT SUPPORT MORE THAN ONE ComponentConfig VERSIONS.
+	// When we need to upgrade the component config version, users will
+	// stop CKE, update cluster.yml in etcd, then start the new CKE.
+	// So, CKE should only support the latest config version.
+	cfg := *base
```

Comment:
Use `DeepCopy()` not to destroy `base`.
Though the current calls of this method seem to be unaffected whether the method is destructive or not, it would be safer to keep methods non-destructive.

Reply:
P.S.
`json.Unmarshal()` is the destroyer.
https://play.golang.org/p/Fo0FxXAltjC